framework/src/main/java/org/checkerframework/framework/util/ComplexQualifierHierarchy.java

```diff
@@ -0,0 +1,380 @@
+        return new QualifierKindHierarchy(qualifierClasses);
+    }
+
+    /**
+     * Creates a mapping from QualifierKind to AnnotationMirror for all qualifiers whose annotations
+     * do not have elements.
+     *
+     * @return the mapping
+     */
+    protected Map<QualifierKind, AnnotationMirror> createQualifiers() {
+        Map<QualifierKind, AnnotationMirror> quals = new TreeMap<>();
+        for (QualifierKind kind : qualifierKindHierarchy.allQualifierKinds()) {
+            if (!kind.hasElements()) {
+                quals.put(kind, AnnotationBuilder.fromClass(elements, kind.getAnnotationClass()));
+            }
+        }
+        return Collections.unmodifiableMap(quals);
+    }
+
+    /**
+     * Creates a mapping from QualifierKind to AnnotationMirror, where the QualifierKind is top.
+     *
+     * <p>Subclasses must override this if the top annotation has elements and provides no default.
+     *
+     * @return a mapping from QualifierKind to AnnotationMirror, where the QualifierKind is top
+     */
+    protected Map<QualifierKind, AnnotationMirror> createTops() {
+        Map<QualifierKind, AnnotationMirror> topsMap = new TreeMap<>();
+        for (QualifierKind kind : qualifierKindHierarchy.getTops()) {
+            topsMap.put(kind, AnnotationBuilder.fromClass(elements, kind.getAnnotationClass()));
+        }
+        return topsMap;
+    }
+
+    /**
+     * Creates a mapping from QualifierKind to AnnotationMirror, where the QualifierKind is bottom.
+     *
+     * <p>Subclasses must override this if the bottom annotation has elements and provides not
+     * default.
+     *
+     * @return a mapping from QualifierKind to AnnotationMirror, where the QualifierKind is bottom
+     */
+    protected Map<QualifierKind, AnnotationMirror> createBottoms() {
+        Map<QualifierKind, AnnotationMirror> bottomsMap = new TreeMap<>();
+        for (QualifierKind kind : qualifierKindHierarchy.getBottoms()) {
+            bottomsMap.put(kind, AnnotationBuilder.fromClass(elements, kind.getAnnotationClass()));
+        }
+        return bottomsMap;
+    }
+
+    /**
+     * Returns the qualifier kind for the given annotation.
+     *
+     * @param anno annotation mirror
+     * @return the qualifier kind for the given annotation
+     */
+    protected QualifierKind getQualifierKind(AnnotationMirror anno) {
+        String name = AnnotationUtils.annotationName(anno);
+        return getQualifierKind(name);
+    }
+
+    /**
+     * Returns the qualifier kind for the annotation with the fully qualified name {@code name}.
+     *
+     * @param name fully qualified annotation name
+     * @return the qualifier kind for the annotation with {@code name}
+     */
+    protected QualifierKind getQualifierKind(String name) {
+        QualifierKind kind = qualifierKindHierarchy.getQualifierKind(name);
+        if (kind == null) {
+            throw new BugInCF("QualifierKind not in hierarchy: %s", name);
```

Comment:
Should this be TypeSystemError?  Similar question about some other uses of BugInCF.

Reply:
No, the methods in this class are called by the Checker Framework during type checking.  So although these could be a bug in the type system, it is just as likely a bug in the Checker Framework.